app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -136,58 +159,105 @@ public void onBtnBackClick() {
+                    Toast.makeText(this, R.string.toast_write_storage_permission_required,
+                            Toast.LENGTH_LONG).show();
+                    return;
+                }
+                break;
+            case API_LEVEL_30:
+                if (!upgradedToApiLevel30()) {
+                    Toast.makeText(this, R.string.toast_api_level_30_must_reset,
+                            Toast.LENGTH_LONG).show();
+                    return;
+                }
+                break;
         }
 
-        int current = getItem(+1);
-        if (current < mLayouts.length) {
-            // Move to next slide.
-            mViewPager.setCurrentItem(current);
-            mBackButton.setVisibility(View.VISIBLE);
-        } else {
+        int next = mViewPager.getCurrentItem() + 1;
+        while (next < slideOrder.length) {
+            if (!shouldSkipSlide(slideOrder[next])) {
+                mViewPager.setCurrentItem(next);
+                mBackButton.setVisibility(View.VISIBLE);
+                break;
+            }
+            next++;
+        }
+        if (next == mLayouts.length) {
             // Start the app after "mNextButton" was hit on the last slide.
             Log.v(TAG, "User completed first start UI.");
             mPreferences.edit().putBoolean(Constants.PREF_FIRST_START, false).apply();
             startApp();
         }
     }
 
-    private void addBottomDots(int currentPage) {
-        mDots = new TextView[mLayouts.length];
+    private boolean isFirstStart() {
+        return mPreferences.getBoolean(Constants.PREF_FIRST_START, true);
+    }
 
-        int[] colorsActive = getResources().getIntArray(R.array.array_dot_active);
-        int[] colorsInactive = getResources().getIntArray(R.array.array_dot_inactive);
+    private boolean upgradedToApiLevel30() {
+        boolean upgraded = mPreferences.getBoolean(Constants.PREF_UPGRADED_TO_API_LEVEL_30,false);
+        if (upgraded) {
+            return true;
+        }
+        if (isFirstStart()) {
+            mPreferences.edit().putBoolean(Constants.PREF_UPGRADED_TO_API_LEVEL_30, true).apply();
+            return true;
+        }
+        return false;
+    }
 
-        mDotsLayout.removeAllViews();
+    private Slide currentSlide() {
+        return slideOrder[mViewPager.getCurrentItem()];
+    }
+
+    private boolean shouldSkipSlide(Slide slide) {
+        switch (slide) {
+            case INTRO:
+                return !isFirstStart();
+            case STORAGE:
+                return PermissionUtil.haveStoragePermission(this);
+            case LOCATION:
+                return hasLocationPermission();
+            case API_LEVEL_30:
+                // Skip if running as root, as that circumvents any Android FS restrictions.
+                return upgradedToApiLevel30()
+                        || PreferenceManager.getDefaultSharedPreferences(this).getBoolean(Constants.PREF_USE_ROOT, false);
```

Comment:
What about a case where I skip over this as I am running as root, then disable root, and everything potentially gets nuked?

Reply:
That's an experimental setting - to me using root means you are ok with anything. I really don't want to handle edge cases around root.